[Linux only] Upgrading to CEF 2704

nethip opened the pull request:
### CEF verison upgrade on Linux to 2704.

Note: CEF version on Windows and Mac will still remain to be 2623. At a later time we will upgrade CEF on all platforms to one version.

The main reason why we could not update CEF on linux to 2171/2623 was because of app quit blockage. Upon App quit, the process was completely blocked as the app was waiting for one of the threads to join. After lot of debugging and analysis, I found out that this was because of Linux sandboxing, `gtk_init()` needs to be called after `CefInitialize()`.

The other major change is the removal of `libcrypto` dependency, users now don't have to install this package for Brackets to run on Linux.

_Here are other noteworthy changes._

1) From now on `cefclient_gtk.cc` will be the entry point for the app on Linux
2) I have modelled the enitre app to that of `CefClient` test app. 
3) We are going to be using `client::ClientHandler` instead of `appshell/ClientHandler`. The difference is that the former will now derive from `appshell/ClientHandler`. 
4) The App quit is now wired to `RootWindowGtk` close. `client::RootWindowGtk` is part of `CefClient`. All issues relating to app quit/ closing individual windows are now fixed.(Yay!!!).
5) Improved multiple windows workflow. `g_handler` is not going to be used anymore as a new client handler is created for every window creation by `RootWindow` itself. This improves our multiple windows workflow. Also `Debug->New Window`, will now have it's own menu bar (finally!). So each of the windows now have their own life time.
6) Menus creation and handling is now mapped to `RootWindowGtk`. `RootWindowGtk` will further deliver specific command to the respective browser.
7) `appshell_extensions_gtk.cpp` had an outdated version on `master` as it was always updated in the `linux-1547` branch alone. So migrated all changes, relating to `appshell_extensions_gtk.cpp` present in `linux-1547` branch to `master`.
8) All the changes are pertaining only to Linux. Wherever required I have wrapped all the code with `#ifdef OS_LINUX` macro
9) On Linux we are going to call `startNodeProcess()` at the start itself. If the thread, that spawns the node process, is called after `CefInitialize()`, it blocks the app quit.
10) Wired in the new developer tools.
11) Have modified `cef_paths.gypi` inside the redistributable for Linux 32 bit, zipped and uploaded that to S3.
12) Made changes to `control` file to remove the `libcrypto` dependency


Note to the committers: Whoever is merging, feel free to squash all the commits and merge.

_**Tasks**_

- [x] Install on Linux 14 and 16 LTS (both 32 and 64) systems and check if Brackets is running fine on these systems
- [x] Verify if libcrypto dependency is gone
- [x] Build this branch on all platforms and see Win and Mac are building fine.
- [x]  Run unit tests
- [ ] Sometimes, there is a performance bug with respect to startup, because of https://github.com/adobe/brackets/blob/master/src/filesystem/FileSystemEntry.js#L194. I could narrow down it down to recursion on startup. I will unit test more on different systems and take a call.

@ficristo, @ingorichter, @eyelash, @zaggino   It would be great if you can review this PR.

ficristo commented:
Just to verify if I understood correctly: are you going to update macOS and Windows with cef 2704 ?
But in a different PR ?
Also, since you dind't use the current cef version used for macOS / Windows (2623) why you didn't use a newer version instead?
I'm not pressing to update it, I only want to know why.

nethip commented:
@ficristo What I meant was that when we upgrade we will upgrade CEF on all the platforms to the latest version but for now only Brackets on Linux would be on 2704. Also I did not upgrade Win and Mac CEF versions to 2704 as I did not want Brackets to be effected on other platforms so kept the changes as minimal as possible and restricted only to Linux platform.

I tried updating to the latest CEF version, but was not successful building Linux atleast. The build mechanism has changed for building CefClient in the most recent versions of CEF(they moved away from gyp). So I guess we need to be update our build mechanism to that of the latest CefClient. We could do this in a later PR.

swmitra commented:
💯 👍 :1st_place_medal:

nethip commented:
@ficristo Will you be able to review this PR and let me know if you have any comments?

ficristo commented:
Sorry, I cannot really review C++ code.
Next week I should be able to, at least, try this PR. I just need to undust my VM...

@pratts you seemed interested in the past in Linux development of Brackets. Do you want to give a try to this PR ?

ficristo commented:
In my light testing I can confirm the crash on quit is fixed.
It is also nice to have the developer tools inside Brackets itself.
I didn't try to build the installer.

I don't know how much it is important but when you scroll an element (the editor area, the file tree or the extension manager) I see the following line printed in the terminal:
`[0805/104322:ERROR:PlatformKeyboardEvent.cpp(117)] Not implemented reached in static PlatformEvent::Modifiers blink::PlatformKeyboardEvent::getCurrentModifierState()`

pratts commented:
Thank you for this update @nethip. I really appreciate the effort you put in !
I cloned the brackets source code and checked out master for brackets and nethip/linux-cef-2704-upgrade for brackets-shell.
Brackets's grunt build and grunt installer worked fine. It no longer asked for libgcrypt11 👍 
Trying to install the .deb file gives the following output:
```
prateek:/home/prateek/brackets$ sudo dpkg -i Brackets\ Release\ 1.11\ 64-bit.deb 
[sudo] password for prateek: 
Selecting previously unselected package brackets.
(Reading database ... 359397 files and directories currently installed.)
Preparing to unpack Brackets Release 1.11 64-bit.deb ...
Unpacking brackets (1.11.0-17496) ...
Setting up brackets (1.11.0-17496) ...
**xdg-icon-resource: size argument must be numeric
Try 'xdg-icon-resource --help' for more information.**
Processing triggers for menu (2.1.47) ...
Processing triggers for hicolor-icon-theme (0.13-1) ...
```

Running brackets for the first time from terminal gives the following error:
brackets 
[0806/133905:FATAL:setuid_sandbox_host.cc(162)] The SUID sandbox helper binary was found, but is not configured correctly. Rather than run without sandboxing I'm aborting now. You need to make sure that /opt/brackets/chrome-sandbox is owned by root and has mode 4755.
#0 0x7fb81b590c3e base::debug::StackTrace::StackTrace()
#1 0x7fb81b5a59fb logging::LogMessage::~LogMessage()
#2 0x7fb81e5a3ad6 sandbox::SetuidSandboxHost::PrependWrapper()
#3 0x7fb81e4ee917 content::ZygoteCommunication::Init()
#4 0x7fb81e4ef31e content::CreateZygote()
#5 0x7fb81e218303 content::BrowserMainLoop::EarlyInitialization()
#6 0x7fb81e21ed79 content::BrowserMainRunnerImpl::Initialize()
#7 0x7fb81b502d19 CefMainDelegate::RunProcess()
#8 0x7fb81e6708e2 content::RunNamedProcessTypeMain()
#9 0x7fb81e671257 content::ContentMainRunnerImpl::Run()
#10 0x7fb81b496e6e CefContext::Initialize()
#11 0x7fb81b496c4b CefInitialize()
#12 0x7fb81b43acf6 cef_initialize
#13 0x0000005e1e38 CefInitialize()
#14 0x000000590c87 client::MainContextImpl::Initialize()
#15 0x000000570e7c client::(anonymous namespace)::RunMain()
#16 0x000000571571 main
#17 0x7fb817037b45 __libc_start_main
#18 0x00000055c779 <unknown>

Aborted

I changed the file permission using chmod 4755 /opt/brackets/chrome-sandbox

One more problem, I cannot see the menu bar :File,Edit,View, etc.

pratts commented:
I can close the window using Ctrl+Q. It gives an error and window closes but I see an exception on console and brackets process does not terminate.

```
brackets
[0806/151520:ERROR:nss_util.cc(842)] After loading Root Certs, loaded==false: NSS error code: -8018
prateek@prateek-debian:~/work/open-source$ /opt/brackets/node-core/ConnectionManager.js:76
            console.error("[Connection] Unable to stringify message: " + e.message);
            ^

RangeError: Maximum call stack size exceeded
    at Connection._send (/opt/brackets/node-core/ConnectionManager.js:76:13)
    at Connection.sendEventMessage (/opt/brackets/node-core/ConnectionManager.js:203:14)
    at /opt/brackets/node-core/ConnectionManager.js:242:11
    at Array.forEach (native)
    at Object.sendEventToAllConnections (/opt/brackets/node-core/ConnectionManager.js:241:18)
    at Object.emitEvent (/opt/brackets/node-core/DomainManager.js:225:27)
    at EventEmitter.<anonymous> (/opt/brackets/node-core/BaseDomain.js:123:28)
    at emitThree (events.js:116:13)
    at EventEmitter.emit (events.js:194:7)
    at logReplacement (/opt/brackets/node-core/Logger.js:80:12)
```

saurabh95 commented on appshell/appshell_extensions.cpp:
```diff
@@ -394,10 +399,15 @@ class ProcessMessageDelegate : public ClientHandler::ProcessMessageDelegate {
+            #endif
+
+            #ifndef OS_LINUX
+                browser->GetHost()->ShowDevTools(wi, browser->GetHost()->GetClient(), settings, CefPoint());
```
This statement seems to be executed in case of MAC, so can we have a separate `#elif` case in line 406 only.
Something like `#ifdef OS_MAC`, it would make it more readable.

nethip replied:
This statement needs to be executed for MAC as well.

nethip replied:
I will see if I can simplify this for better readability.

saurabh95 commented on appshell/appshell_extensions.cpp:
```diff
@@ -765,8 +783,53 @@ class ProcessMessageDelegate : public ClientHandler::ProcessMessageDelegate {
-		}
-		else {
+		} 
+        else if (message_name == "ReadDirWithStats") {
```
Could you please highlight the place where this message is called.

nethip replied:
So here is the thing. In my testing what I figured out was `readDir` leads to lot of shell calls. Here is what happens for `readDir`. First we get plain contents of a directory and for every file entry, we then make a `stat` call to shell. With our Brackets repository open, this led to close to 33,000 shell calls. So if all we want is to get directory contents and file stats we could very well do that in a single shell call. Using this has reduced the no of shell calls to 5000 for readDir, for Brackets repo. So I guess this should pretty much be made mainstream. For now, I have just kept it added and it another PR on the Brackets side, I would replace "ReadDir" shell call with "ReadDirWithStats". Do you think it makes sense to remove this here and have another PR for this?

nethip commented:
Thanks @ficristo @pratts for trying this branch out and testing! Appreciate your efforts 👍 

@ficristo I will dig deeper into it and see which module is reporting this error.

@pratts Thanks for your efforts. I will look into the issues you have mentioned above. I have done all the development on Ubuntu and there I have not seen this error. I will try to introduce the `chmod` through a grunt job. About the other console errors, I think the following is reported by Node side of things. I ran once into this error, but could not repro it again. I will try to  repro this again and see.

```
RangeError: Maximum call stack size exceeded
    at Connection._send (/opt/brackets/node-core/ConnectionManager.js:76:13)
    at Connection.sendEventMessage (/opt/brackets/node-core/ConnectionManager.js:203:14)
    at /opt/brackets/node-core/ConnectionManager.js:242:11
    at Array.forEach (native)
    at Object.sendEventToAllConnections (/opt/brackets/node-core/ConnectionManager.js:241:18)
    at Object.emitEvent (/opt/brackets/node-core/DomainManager.js:225:27)
    at EventEmitter.<anonymous> (/opt/brackets/node-core/BaseDomain.js:123:28)
    at emitThree (events.js:116:13)
    at EventEmitter.emit (events.js:194:7)
    at logReplacement (/opt/brackets/node-core/Logger.js:80:12)

```

And about the Menus not being shown, are they currently shown if you build brackets-shell from linux-1547? Is this happening with this branch alone?

saurabh95 commented on appshell/appshell_extensions_gtk.cpp:
```diff
@@ -72,24 +93,9 @@
 #define KEY_BACK_QUOTE "`"
 #define KEY_COMMA ","
 
+typedef char s8;
```
This is not used anywhere. I guess we can remove it.

nethip replied:
👍

saurabh95 commented on appshell/common/client_app_delegates_common.cc:
```diff
@@ -3,15 +3,16 @@
 
 #include "appshell/common/client_app.h"
-#include "appshell/common/scheme_test_common.h"
+//#include "appshell/common/scheme_test_common.h"
```
nit

saurabh95 commented on appshell/common/client_app_delegates_common.cc:
```diff
@@ -3,15 +3,16 @@
     CefRefPtr<CefSchemeRegistrar> registrar,
     std::vector<CefString>& cookiable_schemes) {
-  scheme_test::RegisterCustomSchemes(registrar, cookiable_schemes);
+  // Brackets specific change.
```
nit

saurabh95 commented on appshell/renderer/client_app_delegates_renderer.cc:
```diff
@@ -4,14 +4,16 @@
 #include "appshell/renderer/client_app_renderer.h"
 #include "appshell/renderer/client_renderer.h"
-#include "appshell/renderer/performance_test.h"
+//#include "appshell/renderer/performance_test.h"
```
nit

saurabh95 commented on appshell/renderer/client_app_delegates_renderer.cc:
```diff
@@ -4,14 +4,16 @@
 void ClientAppRenderer::CreateDelegates(DelegateSet& delegates) {
   renderer::CreateDelegates(delegates);
-  performance_test::CreateDelegates(delegates);
+  // Brackets specific change.
```
nit

pratts commented:
Hi @nethip 
I can see the menu bar in build made from linux-1547 branch

I had to install libgcrypt11 library and saw the following error in terminal:

**[0807/190453:ERROR:connection.cc(799)] sqlite error 11, errno 0: malformed database schema (is_transient) - near "where": syntax error
[0807/190453:ERROR:connection.cc(799)] sqlite error 11, errno 0: malformed database schema (is_transient) - near "where": syntax error
ATTENTION: default value of option force_s3tc_enable overridden by environment.
[0807/190514:ERROR:browser_main_loop.cc(212)] GLib-GObject: invalid cast from 'GtkExpandedContainer' to 'GtkWindow'
[0807/190514:ERROR:browser_main_loop.cc(212)] Gtk: IA__gtk_window_present_with_time: assertion 'GTK_IS_WINDOW (window)' failed**

![screenshot from 2017-08-07 19-05-12](https://user-images.githubusercontent.com/5246231/29028955-94b89920-7ba3-11e7-916d-c6b028949035.png)

ingorichter commented:
@nethip did you ask to test on Ubuntu? Which version of Ubuntu? 16/04, 16/10, 17/04? Older versions or other Linux distros?

nethip commented:
@ingorichter I had done a round of testing on both Ubuntu 14 and 16. But testing on other distros would be very helpful.

nethip commented:
@pratts That is a bummer. I don't have the debian VM. I will set one up. Meanwhile, will you be able to debug the app. I will point you to the sources where menus are getting created, to see which specific call is failing?

pratts commented:
@nethip I use debian 8.
I would love to try this. ❤️

nethip commented:
Awesome @pratts . Thanks for the help. Here is a brief of what I generally do.

1) Enable debug symbols creation in `appshell.gyp`. Add `-g` to `CFLAGS`. [Add it here](https://github.com/adobe/brackets-shell/blob/nethip/linux-cef-2704-upgrade/appshell.gyp#L315)
2) All menu handling is now taken care of by `appshell/browser/root_window_gtk.cc` ([root_window_gtk.cc:528 for creation](https://github.com/adobe/brackets-shell/blob/nethip/linux-cef-2704-upgrade/appshell/browser/root_window_gtk.cc#L528) & [root_window_gtk.cc:56 for menu activation handling](https://github.com/adobe/brackets-shell/blob/nethip/linux-cef-2704-upgrade/appshell/browser/root_window_gtk.cc#L56))
3) We get the `vbox` while dealing with menus over here [appshell_extensions_platform.cpp:145](https://github.com/adobe/brackets-shell/blob/nethip/linux-cef-2704-upgrade/appshell/appshell_extensions_platform.cpp#L145)
4) Menus are created here. [appshell_extensions_gtk.cpp:1042](https://github.com/adobe/brackets-shell/blob/nethip/linux-cef-2704-upgrade/appshell/appshell_extensions_gtk.cpp#L1042)

I am suspecting step 3 is failing. But do let me know, if anything else is breaking.

nethip commented:
@pratts Myself and @saurabh95 had fixed the menus not being shown on Debian 9. Fixed multiple issues with content not getting shown for new window and dev tools workflows. Thanks @saurabh95 for collaborating with me in fixing the bug.

Also changed installer scripts to include `chmod` and `chown` steps for linux installation.

nethip commented:
@pratts Thanks for testing and reporting the bugs. Can you pull the latest revision of this branch and build to see if the reported errors are gone?

pratts commented:
@nethip @saurabh95 Thank you for the quick response 👍 
I checked the previously mentioned issues and did not see them again till now.
I saw something on terminal after launching brackets:
```
[0809/223422:ERROR:CONSOLE(6277)] "Input positions array is not sorted or contains duplicate values.", source: chrome-devtools://devtools/inspector.js (6277)
[0809/223459:ERROR:PlatformKeyboardEvent.cpp(117)] Not implemented reached in static PlatformEvent::Modifiers blink::PlatformKeyboardEvent::getCurrentModifierState()
```
First one after opening the developer tools and second one after opening a file and scrolling to see the content.
Will update if I find any other problems.

nethip commented:
@pratts 
```
[0809/223459:ERROR:PlatformKeyboardEvent.cpp(117)] Not implemented reached in static PlatformEvent::Modifiers blink::PlatformKeyboardEvent::getCurrentModifierState()
```

This looks like a chromium bug. I had looked at the source file and it looks like once we update to the latest CEF, this bugs goes off, but until then we can't do much about it.

Ref: https://bugs.chromium.org/p/chromium/issues/detail?id=538289

saurabh95 commented on installer/linux/build_installer.sh:
```diff
@@ -9,6 +9,8 @@ chmod 755 debian/package-root/opt/brackets/www/node_modules/opn/xdg-open
 chmod 755 debian/package-root/DEBIAN/prerm
 chmod 755 debian/package-root/DEBIAN/postrm
 chmod 755 debian/package-root/DEBIAN/postinst
+chmod 4755 debian/package-root/opt/brackets/chrome-sandbox
```
I think we need to change the order of these commands

nethip commented on installer/linux/build_installer.sh:
```diff
@@ -9,6 +9,9 @@ chmod 755 debian/package-root/opt/brackets/www/node_modules/opn/xdg-open
 chmod 755 debian/package-root/DEBIAN/postrm
 chmod 755 debian/package-root/DEBIAN/postinst
+chown root debian/package-root/opt/brackets/chrome-sandbox
+chmod 4755 debian/package-root/opt/brackets/chrome-sandbox
```
@saurabh95 I have changed the order now.

ingorichter commented:
@nethip I setup up a clean VM with Ubuntu 17.04 64 Bit.

I'm running into this compile error
```
Running "build-linux" task
>> In file included from ./appshell/browser/client_handler.h:19:0,
>>                  from ./appshell/browser/browser_window.h:11,
>>                  from ./appshell/browser/browser_window_std_gtk.h:9,
>>                  from appshell/browser/browser_window_std_gtk.cc:5:
>> ./appshell/browser/dialog_handler_gtk.h:31:8: error: ‘bool client::ClientDialogHandlerGtk::OnJSDialog(CefRefPtr<CefBrowser>, const CefString&, const CefString&, CefJSDialogHandler::JSDialogType, const CefString&, const CefString&, CefRefPtr<CefJSDialogCallback>, bool&)’ marked ‘override’, but does not override
>>    bool OnJSDialog(CefRefPtr<CefBrowser> browser,
>>         ^~~~~~~~~~
>> make: * [out/Release/obj.target/Brackets/appshell/browser/browser_window_std_gtk.o] Error 1
```

The compiler version is `gcc version 6.3.0 20170406`. What is the compiler version you use for building?

nethip commented:
@ingorichter Apologies for delayed response. Just came back from vacation.

Thanks for building this branch and reporting the error. We got similar error messages on Debian 9. I will look into it.

ingorichter commented:
No worries. I wanted to be sure that I didn't miss anything

swmitra commented:
Awesome work @nethip 💯 . This should resolve most of the problems encountered by Brackets users on Linux 👍

ingorichter commented:
@nethip I'll build again today. But it got merged already and therefore I assume it's no longer failing

nethip commented:
@ingorichter We had fixed the errors that you had reported. We were able to run into these errors on Debian and thanks for trying this branch and reporting those.

Will you be able to do some testing around startup performance? I wanted to check if, with the new CEF, the launch time is not effected. Testing involves just loading Brackets repo (with npm install run), re-launching Brackets to see if the startup is slow with bug projects. In my testing I got mixed results, but would be great if you can confirm this.

ingorichter commented:
@nethip That is great. I can look at it tonight or tomorrow

ingorichter commented:
Unfortunately, things get merged faster than I sometimes can test them. :-(
I still have compilation warnings and the linking step didn't finish successfully for me. I'm a bit surprised that nobody saw this issue before.

nethip commented:
@ingorichter Could you post the issues reported?  I will have a look into it. We have fixed the one's you had reported above.

We were getting a linker error on build system to do with libxss1 not being present. If this is the same error, the following command needs to be run on the terminal.

`sudo apt-get install libxss1`

Do let me know if you are getting any other error.

We checked for compilation on multiple systems(ubuntu x86, x64 as well as Debian 8), including build system and posted pre-release builds for the same.

nethip commented:
I am in the process of creating a wiki for building brackets-shell on CEF and I will update the page with all possible errors that one might run into while building and the possible solutions for the same.
